File: 02_Go/worker.go
package main

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"sync"
	"time"
	"workers/incoming"
)

type concurrentSet[T comparable] struct {
	mu sync.Mutex
	m  map[T]struct{}
}

func (s *concurrentSet[T]) tryAdd(v T) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.m[v]; exists {
		return false
	}

	s.m[v] = struct{}{}
	return true
}

var processed = &concurrentSet[uint]{
	m: make(map[uint]struct{}),
}

type Worker struct {
	reader <-chan *incoming.Work
}

func NewWorker(reader <-chan *incoming.Work) *Worker {
	return &Worker{
		reader: reader,
	}
}

func (w *Worker) Execute(ctx context.Context) {
	for {
		select {
		case <-ctx.Done():
			return
		case work, ok := <-w.reader:
			if !ok {
				return
			}

			log.Printf(
				"Working on item %d | Model=%v | Behaviour=%v",
				work.GetId(),
				work.GetExecutionModel(),
				work.GetBehaviour())

			w.simulate(work, ctx)

			log.Printf("Done work %d.", work.GetId())
		}
	}
}

func (w *Worker) simulate(work *incoming.Work, ctx context.Context) {
	behaviour := work.GetBehaviour()
	// Exactly-once: skip duplicates (demo-level, in-memory only).
	if behaviour.HasFlag(incoming.BehaviourExactlyOnce) && !processed.tryAdd(work.GetId()) {
		log.Printf("  [ExactlyOnce] Work %d already processed -> skipping.", work.GetId())
		return
	}

	var attempts int
	if behaviour.HasFlag(incoming.BehaviourRetryable) {
		attempts = 3
	} else {
		attempts = 1
	}

	for attempt := 1; attempt <= attempts; attempt++ {
		select {
		case <-ctx.Done():
			log.Printf("  Work cancelled")
			return
		default:
			// HighPriority: less "queueing"/overhead before work begins.
			if behaviour.HasFlag(incoming.BehaviourHighPriority) {
				log.Printf("  [HighPriority] Fast-lane execution.")
			}

			// RequiresAffinity: pin to a pretend partition/worker lane.
			lane := work.GetId() % 4
			if behaviour.HasFlag(incoming.BehaviourRequiresAffinity) {
				log.Printf("  [RequiresAffinity] Routing to lane %d.", lane)
			}

			if err := w.simulateByExecutionModel(work, lane); err != nil {
				log.Printf("  Attempt %d/%d failed: %v", attempt, attempts, err)

				if attempt == attempts {
					log.Printf("  Giving up.")
					return
				}

				// Tiny exponential-ish backoff for the demo.
				backoff := time.Duration(50 * attempt * attempt)
				time.Sleep(backoff * time.Millisecond)
				continue
			}

			if behaviour.HasFlag(incoming.BehaviourRetryable) {
				log.Printf("  [Retryable] Succeeded on attempt %d/%d.", attempt, attempts)
			} else {
				log.Printf("  Completed.")
			}
		}
	}
}

func (w *Worker) simulateByExecutionModel(work *incoming.Work, lane uint) error {
	behaviour := work.GetBehaviour()

	// Base pacing. Behaviours can tweak it.
	stepDelayMs := 80

	if behaviour.HasFlag(incoming.BehaviourLongRunning) {
		stepDelayMs += 160
		log.Printf("  [LongRunning] Slower steps.")
	}

	if behaviour.HasFlag(incoming.BehaviourResourceIntensive) {
		log.Printf("  [ResourceIntensive] Adding CPU work.")
	}

	// Scheduled: pretend we had to wait until a trigger time.
	if work.GetExecutionModel() == incoming.ExecutionModelScheduled {
		log.Printf("  [Scheduled] Waiting for trigger...")
		time.Sleep(150 * time.Millisecond)
	}

	switch work.GetExecutionModel() {
	case incoming.ExecutionModelOneOff:
		if err := step("OneOff: run once", behaviour, stepDelayMs); err != nil {
			return err
		}
		break
	case incoming.ExecutionModelScheduled:
		if err := step("Scheduled: execute job", behaviour, stepDelayMs); err != nil {
			return err
		}
		break
	case incoming.ExecutionModelEventDriven:
		if err := step("EventDriven: handle event payload", behaviour, stepDelayMs); err != nil {
			return err
		}
		break
	case incoming.ExecutionModelBatch:
		log.Printf("  [Batch] Processing items...")
		for i := 1; i <= 5; i++ {
			if err := step(fmt.Sprintf("Batch item %d/5", i), behaviour, stepDelayMs); err != nil {
				return err
			}
		}
		break
	case incoming.ExecutionModelStream:
		log.Printf("  [Stream] Polling/consuming stream ticks...")
		for tick := 1; tick <= 4; tick++ {
			if err := step(fmt.Sprintf("Stream tick %d/5", tick), behaviour, stepDelayMs+40); err != nil {
				return err
			}
		}
		break
	case incoming.ExecutionModelWorkflow:
		log.Printf("  [Workflow] Running steps (DAG-ish)...")
		if err := step("Step A: validate", behaviour, stepDelayMs); err != nil {
			return err
		}
		if err := step("Step B: transform", behaviour, stepDelayMs+20); err != nil {
			return err
		}
		if err := step("Step C: persist", behaviour, stepDelayMs+40); err != nil {
			return err
		}
		break
	case incoming.ExecutionModelActor:
		log.Printf("  [Actor] Handling partition key lane=%d (stateful-ish).", lane)
		if err := step("Actor turn: load state", behaviour, stepDelayMs); err != nil {
			return err
		}
		if err := step("Actor turn: apply work", behaviour, stepDelayMs+20); err != nil {
			return err
		}
		if err := step("Actor turn: save state", behaviour, stepDelayMs+40); err != nil {
			return err
		}
		break
	default:
		if err := step("Unknown model: fallback", behaviour, stepDelayMs); err != nil {
			return err
		}
		break
	}

	return nil
}

func step(label string, behaviour incoming.Behaviour, delayMs int) error {
	log.Printf("  -> %s", label)
	if behaviour.HasFlag(incoming.BehaviourResourceIntensive) {
		cpuBump(25_000)
	}

	if err := maybeFailTransiently(behaviour); err != nil {
		return err
	}

	time.Sleep(time.Duration(delayMs) * time.Millisecond)
	return nil
}

// ResourceIntensive: do a tiny CPU spin per step (demo only).
func cpuBump(iterations int) {
	start := time.Now()

	x := 0
	for i := 0; i < iterations; i++ {
		x = (x * 31) ^ i
	}

	_ = x // prevent optimization

	_ = time.Since(start)
}

// Random transient failure for Retryable demos.
func maybeFailTransiently(behaviour incoming.Behaviour) error {
	if !behaviour.HasFlag(incoming.BehaviourRetryable) {
		return nil
	}

	// Fail sometimes (more often when "resource intensive") so retries are visible.
	var odds int // 1/4 or 1/7
	if behaviour.HasFlag(incoming.BehaviourResourceIntensive) {
		odds = 4
	} else {
		odds = 7
	}

	if rand.Intn(odds) == 0 {
		return fmt.Errorf("transient failure (simulated)")
	}

	return nil
}

File: 05_Elixir/workers/lib/worker.ex
defmodule Worker do
  alias Incoming.ExecutionModel
  alias Extras.ExactlyOnceLedger
  alias Incoming.Behaviour
  alias Extras.BoundedChannel

  def start_link(channel) do
    Task.start_link(fn -> execute(channel) end)
  end

  def execute(channel) do
    case BoundedChannel.read(channel) do
      {:ok, work} ->
        IO.puts(
          "Working on item #{work.id} | Model=#{work.execution_model} | Behaviour=#{Behaviour.to_string(work.behaviour)}"
        )

        simulate(work)

        IO.puts("Done work #{work.id}.")

        execute(channel)

      :closed ->
        :ok
    end
  end

  defp simulate(work) do
    # Exactly-once: skip duplicates (demo-level, in-memory only).
    if Behaviour.has?(work.behaviour, Behaviour.exactly_once()) and
         not ExactlyOnceLedger.try_add(work.id) do
      IO.puts("  [ExactlyOnce] Work #{work.id} already processed -> skipping.")
    else
      attempts =
        if Behaviour.has?(work.behaviour, Behaviour.retryable()) do
          3
        else
          1
        end

      for attempt <- 1..attempts do
        try do
          # HighPriority: less "queueing"/overhead before work begins.
          if Behaviour.has?(work.behaviour, Behaviour.high_priority()) do
            IO.puts("  [HighPriority] Fast-lane execution.")
          end

          # RequiresAffinity: pin to a pretend partition/worker lane.
          lane = rem(abs(work.id), 4)

          if Behaviour.has?(work.behaviour, Behaviour.requires_affinity()) do
            IO.puts("  [RequiresAffinity] Routing to lane #{lane}.")
          end

          simulate_by_execution_model(work, lane)

          if Behaviour.has?(work.behaviour, Behaviour.retryable()) do
            IO.puts("  [Retryable] Succeeded on attempt #{attempt}/#{attempts}.")
          else
            IO.puts("  Completed.")
          end

          :ok
        rescue
          e ->
            IO.puts("  Attempt #{attempt}/#{attempts} failed: #{Exception.message(e)}")

            if attempt >= attempts do
              IO.puts("  Giving up.")
              :ok
            else
              # Tiny exponential-ish backoff for the demo.
              backoff_ms = 50 * attempt * attempt
              Process.sleep(backoff_ms)
            end
        catch
          :exit, :shutdown ->
            IO.puts("  Work cancelled")

          kind, reason ->
            IO.puts("  Unexpected #{kind}: #{inspect(reason)}")
        end
      end
    end
  end

  defp simulate_by_execution_model(work, lane) do
    # Base pacing. Behaviours can tweak it.
    base_delay_ms = 80

    step_delay_ms =
      if Behaviour.has?(work.behaviour, Behaviour.long_running()) do
        IO.puts("  [LongRunning] Slower steps.")
        base_delay_ms + 160
      else
        base_delay_ms
      end

    if Behaviour.has?(work.behaviour, Behaviour.resource_intensive()) do
      IO.puts("  [ResourceIntensive] Adding CPU work.")
    end

    # Scheduled: pretend we had to wait until a trigger time.
    if work.execution_model == ExecutionModel.scheduled() do
      IO.puts("  [Scheduled] Waiting for trigger...")
      Process.sleep(150)
    end

    case work.execution_model do
      :one_off ->
        step(work, "OneOff: run once", step_delay_ms)

      :event_driven ->
        step(work, "EventDriven: handle event payload", step_delay_ms)

      :batch ->
        IO.puts("  [Batch] Processing items...")

        Enum.each(1..5, fn i ->
          step(work, "Batch item #{i}/5", step_delay_ms)
        end)

      :stream ->
        IO.puts("  [Stream] Polling/consuming stream ticks...")

        Enum.each(1..4, fn tick ->
          step(work, "Stream tick #{tick}/4", step_delay_ms + 40)
        end)

      :workflow ->
        IO.puts("  [Workflow] Running steps (DAG-ish)...")
        step(work, "Step A: validate", step_delay_ms)
        step(work, "Step B: transform", step_delay_ms + 20)
        step(work, "Step C: persist", step_delay_ms + 40)

      :actor ->
        IO.puts("  [Actor] Handling partition key lane=#{lane} (stateful-ish).")
        step(work, "Actor turn: load state", step_delay_ms)
        step(work, "Actor turn: apply work", step_delay_ms + 20)
        step(work, "Actor turn: save state", step_delay_ms + 40)

      :scheduled ->
        step(work, "Scheduled: execute job", step_delay_ms)

      _other ->
        step(work, "Unknown model: fallback", step_delay_ms)
    end
  end

  # Random transient failure for Retryable demos.
  defp maybe_fail_transiently(work) do
    if Behaviour.has?(work.behaviour, Behaviour.retryable()) do
      # Fail sometimes (more often when "resource intensive") so retries are visible.
      # 1/4 or 1/7
      odds =
        if Behaviour.has?(work.behaviour, Behaviour.resource_intensive()) do
          4
        else
          7
        end

      if Enum.random(1..odds) == 0 do
        raise "Transient failure (simulated)."
      end
    end
  end

  # ResourceIntensive: do a tiny CPU spin per step (demo only).
  defp cpu_bump(iterations) do
    {_time_us, _result} =
      :timer.tc(fn ->
        Enum.reduce(0..(iterations - 1), 0, fn i, x ->
          Bitwise.bxor(x * 31, i)
        end)
      end)
  end

  defp step(work, label, delay_ms) do
    IO.puts("  -> #{label}")

    if Behaviour.has?(work.behaviour, Behaviour.resource_intensive()) do
      cpu_bump(25000)
    end

    maybe_fail_transiently(work)
    Process.sleep(delay_ms)
  end
end
